skfda/preprocessing/dim_reduction/_pace.py

```diff
@@ -0,0 +1,1441 @@
+        weighted_diag = (weights * win)[
+            np.arange(weights.shape[1]),
+            np.arange(weights.shape[1]),
+        ]
+        return cast("NDArrayFloat", weighted_diag)
+
+    def _build_rotated_design_matrix(
+        self,
+        t_pairs: NDArrayFloat,
+        r_t_eval: NDArrayFloat,
+    ) -> NDArrayFloat:
+        """Build design matrix for rotated coordinates."""
+        n_eval = r_t_eval.shape[0]
+        n_obs = t_pairs.shape[0]
+        x = np.ones((n_eval, n_obs, 3))
+
+        for i in range(n_eval):
+            delta_r0 = t_pairs[:, 0, 0] - r_t_eval[i, 0]
+            delta_r1 = t_pairs[:, 1, 0] - r_t_eval[i, 1]
+
+            x[i, :, 1] = delta_r0**2
+            x[i, :, 2] = delta_r1
+
+        return x
+
+    def _rotated_cov_lls(
+        self,
+        h: float,
+        r_eval: NDArrayFloat,
+        s_eval: NDArrayFloat,
+        cov_coords: NDArrayFloat,
+        cov_values: NDArrayFloat,
+        win: NDArrayFloat,
+    ) -> NDArrayFloat:
+        """
+        Local linear smoother for rotated covariance estimation.
+
+        Args:
+            h: Bandwidth for the kernel.
+            r_eval: First array of query points where smoother is evaluated.
+            s_eval: Second array of query points where smoother is evaluated.
+            cov_coords: Coordinates of the covariance.
+            cov_values: Values of the covariance.
+            win: Weights for the covariance.
+
+        Returns:
+            n_grid_points x n_grid_points array of smoothed covariance values.
+        """
+        r_cov_coords, r_t_eval = self._rotate_coordinates(cov_coords, r_eval, s_eval)
+
+        active = np.nonzero(win)[0]
+        t_pairs = r_cov_coords[active]
+        cov_values = cov_values[active]
+        win = win[active]
+
+        weights = self._compute_rotated_cov_weights(h, t_pairs, r_t_eval, win)
+        design_matrix = self._build_rotated_design_matrix(t_pairs, r_t_eval)
+
+        x_t = np.transpose(design_matrix, (0, 2, 1))
+        xtw = x_t * weights[:, None, :]
+        xtwx = xtw @ design_matrix
+        xtwy = xtw @ cov_values
+
+        beta = np.linalg.pinv(xtwx) @ xtwy
+
+        return np.array(beta[:, 0])
+
+    def fit(
+        self,
+        X: FDataIrregular,
+        y: object = None,  # noqa: ARG002
+    ) -> PACE:
+        """
+        Compute the ``n_components`` first principal components and saves them.
+
+        Args:
+            X: The functional data object to be analysed.
+            y: Ignored. Only present because of fit function convention.
+
+        Returns:
+            self
+        """
+        # Check that the number of components is smaller than the sample size
+        if self.n_components > len(X.start_indices):
+            error_msg = (
+                "The sample size must be bigger than the number of components",
+            )
+            raise AttributeError(error_msg)
+
+        if self.boundary_effect_interval == (0.0, 1.0):  # noqa: WPS358
```

Comment:
I would just slice always and simplify the code. Let the slicing method optimize that if needed.